### ISSUE TITLE
SWI-Prolog update from 9.2.2 to 9.2.3 broke Herculog - a solution

### ISSUE THREAD

**Issue by wldwrk** [anbre/prolog-jupyter-kernel#18]
Hi,

The SWI-Prolog update from 9.2.2 to 9.2.3 broke Herculog. A solution could be to stop using the SWI-Prolog built in variable binding predicates (expand_query and expand_answer) and use the jupyter_variable_bindings for both SICStus and SWI. I tried this and works with SWI-Prolog 9.2.3, 9.2.4, 9.3.4 with Windows and Ubuntu 22.04.

The SWI-Prolog change log reads [Mar 4 2024]:
ENHANCED: Answer/query expansion. Toplevel variable ($Var) handing is now always handled, regardless of the expand_query/4 or expand_answer/2 hooks. The expand_answer/2 hook is superseded by expand_answer/3.

Please also note that SWI-Prolog does not support the toplevel predicates for common use, but treats them as system internals (a private API). In a blog they write ¨If you want to hack around, simply call toplevel_variables:toplevel_var/2. Don’t blame anyone if it stops working without notice though as it is a private API."

The amendments required to fix Herculog would be:
jupyter_variable_bindings.pl - use this code for both SICStus and SWI. Amendment would be removal of a SWI conditional.
jupyter_term_handling.pl - amend the comments to reflect the amended situation.

The Readme file could be amended to reflect the fix.
A new Pypi distribution would be required for distribution of the amended code.

I have made the above amendments in a local branch on my PC, but have not the proper authorization for a pull request in the hhu-stups - prolog-jupyter-kernel repository, according to GitHub.

Please advise if and how I could assist you with sharing the proposed amendments.

Kind regards wldwrk

**Comment by leuschel** [comment]
Thanks for your comments and efforts. I will try to make it possible for you to generate pull requests.

**Comment by dgelessus** [comment]
Hello, there is no special permission needed to open a pull request. You were able to create an issue, so GitHub should also allow you to create a pull request.

Where exactly do you get a permission error? Are you perhaps trying to push directly to the main repository (https://github.com/hhu-stups/prolog-jupyter-kernel.git)? This will not work, because you are not a member of our repository. But you can create a fork of our repo, push your branch to your fork, and then open a pull request from there.

**Comment by wldwrk** [comment]
Hi Leuschel and Dgelessus,
Thanks for your responses.
The permission error "...it looks like you don´t have write access to hhu-stups/prolog-jupyter-kernel..." appeared when I tried to make a pull request from the GitHub desktop app I'm using. I think dgelessus is right that this would be a push to the main repository. So I followed dgelessus guidance and created a fork, pushed my branch, and opened a pull request to propose my amendments. I then closed the branch. I don't know if this has any consequence for the pull request. I didn´t delete anything so I believe I can re-open if needed.
